mysite/chat/api.py

```diff
@@ -0,0 +1,284 @@
+import injections
+from django.utils import timezone
+from django.shortcuts import get_object_or_404
+from rest_framework.parsers import JSONParser
+from rest_framework import viewsets, generics
+from rest_framework.response import Response
+from rest_framework.authentication import SessionAuthentication
+from rest_framework.permissions import IsAuthenticated
+
+from .models import Message, Chat, ChatDivider
+from .serializers import (
+    MessageSerializer,
+    ChatHistorySerializer,
+    ChatProgressSerializer,
+    ChatResourcesSerializer
+)
+from .services import ProgressHandler, FsmHandler
+from .permissions import IsOwner
+from ct.models import Response as StudentResponse
+from ct.models import UnitLesson
+
+
+inj_alternative = injections.Container()
+inj_alternative['next_handler'] = FsmHandler()
+MessageSerializer = inj_alternative.inject(MessageSerializer)
+
+
+def get_additional_messages(response, chat):
+    student_errors = response.studenterror_set.all()
+    for each in student_errors:
+        map(lambda ul: Message.objects.get_or_create(contenttype='unitlesson',
+                                                     content_id=ul.id,
+                                                     chat=chat,
+                                                     owner=chat.user,
+                                                     input_type='custom',
+                                                     student_error=each,
+                                                     kind='message',
+                                                     is_additional=True),
+            each.errorModel.get_em_resolutions()[1])
+
+
+@injections.has
+class MessagesView(generics.RetrieveUpdateAPIView, viewsets.GenericViewSet):
+    """
+    GET or UPDATE one message.
+    """
+    parser_classes = (JSONParser,)
+    next_handler = injections.depends(ProgressHandler)
+
+    serializer_class = MessageSerializer
+    queryset = Message.objects.all()
+    authentication_classes = (SessionAuthentication,)
+    permission_classes = (IsAuthenticated, IsOwner)
+
+    def retrieve(self, request, *args, **kwargs):
+        message = self.get_object()
+        chat_id = self.request.GET.get('chat_id')
+        if not chat_id:
+            return Response({'errors': 'Request should include chat_id.'})
+        chat = Chat.objects.filter(id=chat_id).first()
+        if not chat:
+            return Response({'errors': 'There is no chat by chat_id.'})
```

Comment:
The following lines of codes are repeated serveral places in this PR. Could it be written in a common util function?
```python
        chat_id = self.request.GET.get('chat_id')
        if not chat_id:
            return Response({'errors': 'Request should include chat_id.'})
        chat = Chat.objects.filter(id=chat_id).first()
        if not chat:
            return Response({'errors': 'There is no chat by chat_id.'})
 ```